analyzers/src/SonarAnalyzer.CSharp/SymbolicExecution/Roslyn/PublicMethodArgumentsShouldBeCheckedForNull.cs

```diff
@@ -29,5 +29,81 @@ public class PublicMethodArgumentsShouldBeCheckedForNull : SymbolicRuleCheck
+            node is BaseMethodDeclarationSyntax { ParameterList.Parameters.Count: > 0 } method
+            && MethodDereferencesArguments(method);
+
+        static bool IsRelevantPropertyAccessor(SyntaxNode node) =>
+            node is AccessorDeclarationSyntax accessor
+            && IsPropertyAccessorAccessibleFromOtherAssemblies(accessor.Modifiers);
+
+        static bool IsAccessibleFromOtherAssemblies(SyntaxNode node) =>
+            node.AncestorsAndSelf().OfType<MemberDeclarationSyntax>().FirstOrDefault() is { } containingMember
+            && node.Ancestors().OfType<BaseTypeDeclarationSyntax>().FirstOrDefault() is { } containingType
+            && IsMemberAccessibleFromOtherAssemblies(containingMember.Modifiers(), containingType)
+            && IsTypeAccessibleFromOtherAssemblies(containingType.Modifiers);
+
+        static bool IsMemberAccessibleFromOtherAssemblies(SyntaxTokenList modifiers, BaseTypeDeclarationSyntax containingType) =>
+            IsPublicOrProtectedOrProtectedInternal(modifiers)
+            || (containingType is InterfaceDeclarationSyntax && HasNoDeclaredAccessabilityModifier(modifiers));
+
+        static bool IsPropertyAccessorAccessibleFromOtherAssemblies(SyntaxTokenList modifiers) =>
+            IsPublicOrProtectedOrProtectedInternal(modifiers)
+            || HasNoDeclaredAccessabilityModifier(modifiers);
+
+        static bool IsTypeAccessibleFromOtherAssemblies(SyntaxTokenList modifiers) =>
+            IsPublicOrProtectedOrProtectedInternal(modifiers);
+
+        static bool IsPublicOrProtectedOrProtectedInternal(SyntaxTokenList modifiers) =>
+            modifiers.AnyOfKind(SyntaxKind.PublicKeyword)
+            || (modifiers.AnyOfKind(SyntaxKind.ProtectedKeyword) && !modifiers.AnyOfKind(SyntaxKind.PrivateKeyword));
+
+        static bool HasNoDeclaredAccessabilityModifier(SyntaxTokenList modifiers) =>
+            !modifiers.Any(x => x.IsAnyKind(
+                SyntaxKind.PrivateKeyword,
+                SyntaxKind.ProtectedKeyword,
+                SyntaxKind.InternalKeyword,
+                SyntaxKind.PublicKeyword));
+
+        static bool MethodDereferencesArguments(BaseMethodDeclarationSyntax method)
+        {
+            var argumentNames = method.ParameterList.Parameters.Select(x => x.GetName()).ToArray();
+            var walker = new ArgumentDereferenceWalker(argumentNames);
+            walker.SafeVisit(method);
+            return walker.DereferencesMethodArguments;
+        }
+    }
+
+    private sealed class ArgumentDereferenceWalker : SafeCSharpSyntaxWalker
+    {
+        private readonly string[] argumentNames;
+
+        public bool DereferencesMethodArguments { get; private set; }
+
+        public ArgumentDereferenceWalker(string[] argumentNames) =>
+            this.argumentNames = argumentNames;
+
+        public override void Visit(SyntaxNode node)
+        {
+            if (!DereferencesMethodArguments)
+            {
+                base.Visit(node);
```

Comment:
Don't visit 
* static local functions
* static lambdas

Because parameters are not accessible there.